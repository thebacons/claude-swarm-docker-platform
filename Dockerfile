# Claude Swarm Docker Container
FROM ubuntu:22.04

# Prevent interactive prompts during installation
ENV DEBIAN_FRONTEND=noninteractive

# Install system dependencies
RUN apt-get update && apt-get install -y \
    curl \
    git \
    python3 \
    python3-pip \
    sudo \
    wget \
    build-essential \
    software-properties-common \
    ca-certificates \
    gnupg \
    lsb-release \
    && rm -rf /var/lib/apt/lists/*

# Install Node.js 20.x
RUN curl -fsSL https://deb.nodesource.com/setup_20.x | bash - \
    && apt-get install -y nodejs \
    && npm install -g npm@latest

# Install Ruby 3.x
RUN apt-get update && apt-get install -y \
    ruby-full \
    ruby-dev \
    && gem install bundler

# Note: Claude Code CLI must be installed manually as it's not available via npm
# The container will need Claude Code installed via the official method

# Create non-root user
RUN useradd -m -s /bin/bash developer \
    && echo "developer ALL=(ALL) NOPASSWD:ALL" >> /etc/sudoers

# Set up workspace
WORKDIR /workspace
RUN mkdir -p projects configs logs sessions \
    && chown -R developer:developer /workspace

# Switch to non-root user
USER developer

# Set environment variables
ENV PATH="/home/developer/.local/bin:${PATH}"
ENV CLAUDE_HOME="/home/developer/.claude"

# Copy configuration files
COPY --chown=developer:developer configs/ /workspace/configs/
COPY --chown=developer:developer *.sh /workspace/

# Note: Scripts will be made executable at runtime

# Default command
CMD ["/bin/bash"]