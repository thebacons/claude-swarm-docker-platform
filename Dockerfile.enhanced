# Claude Swarm Docker Container - Enhanced with Claude Code
FROM ubuntu:22.04

# Build arguments for version management
ARG CLAUDE_CODE_VERSION="latest"
ARG NODE_VERSION="20"
ARG PYTHON_VERSION="3.11"

# Prevent interactive prompts during installation
ENV DEBIAN_FRONTEND=noninteractive

# Install system dependencies
RUN apt-get update && apt-get install -y \
    curl \
    git \
    wget \
    sudo \
    build-essential \
    software-properties-common \
    ca-certificates \
    gnupg \
    lsb-release \
    jq \
    unzip \
    && rm -rf /var/lib/apt/lists/*

# Add Python PPA and install Python 3.11
RUN add-apt-repository ppa:deadsnakes/ppa -y && \
    apt-get update && \
    apt-get install -y \
    python${PYTHON_VERSION} \
    python${PYTHON_VERSION}-venv \
    python${PYTHON_VERSION}-dev \
    python${PYTHON_VERSION}-distutils \
    && rm -rf /var/lib/apt/lists/*

# Set Python 3.11 as default
RUN update-alternatives --install /usr/bin/python3 python3 /usr/bin/python${PYTHON_VERSION} 1 && \
    update-alternatives --install /usr/bin/python python /usr/bin/python${PYTHON_VERSION} 1

# Install pip for Python 3.11
RUN curl -sS https://bootstrap.pypa.io/get-pip.py | python${PYTHON_VERSION}

# Install Node.js
RUN curl -fsSL https://deb.nodesource.com/setup_${NODE_VERSION}.x | bash - && \
    apt-get install -y nodejs && \
    npm install -g npm@latest

# Install Ruby (for potential future use)
RUN apt-get update && apt-get install -y \
    ruby-full \
    ruby-dev \
    && gem install bundler \
    && rm -rf /var/lib/apt/lists/*

# Create non-root user
RUN useradd -m -s /bin/bash developer && \
    echo "developer ALL=(ALL) NOPASSWD:ALL" >> /etc/sudoers

# Set up workspace structure
WORKDIR /workspace
RUN mkdir -p \
    projects \
    configs \
    logs \
    sessions \
    hooks \
    .claude-code \
    scripts \
    && chown -R developer:developer /workspace

# Switch to non-root user for Claude Code installation
USER developer

# Set environment variables
ENV PATH="/home/developer/.local/bin:${PATH}"
ENV CLAUDE_HOME="/home/developer/.claude"
ENV ANTHROPIC_API_KEY=""

# Install Python packages as user (including swarm orchestration dependencies)
RUN pip3 install --user \
    anthropic \
    requests \
    pyyaml \
    python-dotenv \
    redis \
    psycopg2-binary \
    aiofiles \
    asyncio

# Create Claude Code installation script
RUN cat > /workspace/scripts/install-claude-code.sh << 'EOF'
#!/bin/bash
set -e

echo "Installing Claude Code..."

# Check if Claude Code is already installed
if command -v claude 2>&1 >/dev/null; then
    echo "Claude Code is already installed"
    claude --version
    exit 0
fi

# Install Claude Code via npm (when available)
# For now, we'll prepare the environment for manual installation
echo "Claude Code installation prepared."
echo "Please install Claude Code manually following official instructions."

# Create claude wrapper script using Python SDK
cat > /home/developer/.local/bin/claude-sdk << 'WRAPPER'
#!/usr/bin/env python3
import os
import sys
import json
from anthropic import Anthropic

client = Anthropic(api_key=os.environ.get("ANTHROPIC_API_KEY"))

def main():
    if len(sys.argv) < 2:
        print("Usage: claude-sdk <prompt>")
        sys.exit(1)
    
    prompt = " ".join(sys.argv[1:])
    
    try:
        message = client.messages.create(
            model="claude-3-5-sonnet-20241022",
            max_tokens=4096,
            messages=[{"role": "user", "content": prompt}]
        )
        print(message.content[0].text)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    main()
WRAPPER

chmod +x /home/developer/.local/bin/claude-sdk
echo "Claude SDK wrapper installed at /home/developer/.local/bin/claude-sdk"
EOF

RUN chmod +x /workspace/scripts/install-claude-code.sh

# Create version check script
RUN cat > /workspace/scripts/check-claude-version.sh << 'EOF'
#!/bin/bash

echo "=== Claude Code Version Check ==="
echo ""

# Check if Claude Code is installed
if command -v claude 2>&1 >/dev/null; then
    echo "Claude Code version:"
    claude --version 2>/dev/null || echo "Version command not available"
else
    echo "Claude Code: Not installed"
    echo "Using Claude SDK wrapper instead"
fi

echo ""
echo "Python version:"
python3 --version

echo ""
echo "Node.js version:"
node --version

echo ""
echo "Anthropic SDK version:"
pip show anthropic | grep Version || echo "Not installed"

echo ""
echo "=== Environment ==="
echo "ANTHROPIC_API_KEY: ${ANTHROPIC_API_KEY:+[SET]}"
echo "CLAUDE_HOME: $CLAUDE_HOME"
echo ""
EOF

RUN chmod +x /workspace/scripts/check-claude-version.sh

# Create startup script
RUN cat > /workspace/scripts/startup.sh << 'EOF'
#!/bin/bash

echo "=== Claude Swarm Container Starting ==="
echo ""

# Load environment variables from .env if it exists
if [ -f /workspace/.env ]; then
    echo "Loading environment from .env..."
    set -a
    source /workspace/.env
    set +a
fi

# Check versions
/workspace/scripts/check-claude-version.sh

# Install/update Claude Code if requested
if [ "$AUTO_UPDATE_CLAUDE" = "true" ]; then
    echo "Checking for Claude Code updates..."
    /workspace/scripts/install-claude-code.sh
fi

# Set up hooks if directory is mounted
if [ -d /workspace/hooks ]; then
    echo "Setting up Claude Code hooks..."
    mkdir -p /home/developer/.claude-code
    
    # Create settings.json for hooks
    cat > /home/developer/.claude-code/settings.json << 'SETTINGS'
{
  "hooks": {
    "preToolUse": [],
    "postToolUse": [
      {
        "matcher": {
          "tools": ["Write"],
          "files": ["*.js", "*.jsx", "*.ts", "*.tsx"]
        },
        "command": "bash /workspace/hooks/validators/syntax-check.sh ${CLAUDE_FILE_PATH}",
        "blocking": true
      },
      {
        "matcher": {
          "tools": ["Write"],
          "files": ["*.js", "*.jsx"]
        },
        "command": "bash /workspace/hooks/validators/react-check.sh ${CLAUDE_FILE_PATH}",
        "blocking": true
      }
    ],
    "postConversation": []
  }
}
SETTINGS
    echo "Hooks configured."
fi

# Run health check
/workspace/scripts/health-check.sh

echo ""
echo "=== Container Ready ==="
echo ""

# Keep container running
exec "$@"
EOF

RUN chmod +x /workspace/scripts/startup.sh

# Create health check script
RUN cat > /workspace/scripts/health-check.sh << 'EOF'
#!/bin/bash

echo "=== Health Check ==="

# Check API key
if [ -z "$ANTHROPIC_API_KEY" ]; then
    echo "WARNING: ANTHROPIC_API_KEY not set"
    exit 1
fi

# Test Python SDK
python3 -c "import anthropic; print('✓ Anthropic SDK available')" 2>/dev/null || echo "✗ Anthropic SDK error"

# Test network
curl -s -o /dev/null -w "✓ Network connectivity OK\n" https://api.anthropic.com 2>/dev/null || echo "✗ Network error"

# Test file permissions
touch /workspace/test-write 2>/dev/null && rm /workspace/test-write && echo "✓ Write permissions OK" || echo "✗ Write permission error"

echo "=== Health Check Complete ==="
EOF

RUN chmod +x /workspace/scripts/health-check.sh

# Copy configuration files
COPY --chown=developer:developer configs/ /workspace/configs/
COPY --chown=developer:developer hooks/ /workspace/hooks/
COPY --chown=developer:developer *.sh /workspace/

# Create .bashrc additions
RUN cat >> /home/developer/.bashrc << 'EOF'

# Claude Swarm Environment
export ANTHROPIC_API_KEY="${ANTHROPIC_API_KEY}"
export CLAUDE_HOME="/home/developer/.claude"

# Aliases
alias claude-test="python3 -c 'from anthropic import Anthropic; c=Anthropic(); print(c.messages.create(model=\"claude-3-5-sonnet-20241022\", max_tokens=100, messages=[{\"role\": \"user\", \"content\": \"Say hello\"}]).content[0].text)'"
alias claude-chat="python3 /workspace/claude-api.py"
alias check-version="/workspace/scripts/check-claude-version.sh"
alias health-check="/workspace/scripts/health-check.sh"

# Load .env on shell start
if [ -f /workspace/.env ]; then
    set -a
    source /workspace/.env
    set +a
fi

# Welcome message
echo "Welcome to Claude Swarm Docker Container!"
echo "Run 'check-version' to see installed versions"
echo "Run 'health-check' to verify setup"
echo ""
EOF

# Switch back to root for final setup
USER root

# Expose port for potential web UI
EXPOSE 8080

# Health check
HEALTHCHECK --interval=30s --timeout=10s --start-period=5s --retries=3 \
    CMD su - developer -c "/workspace/scripts/health-check.sh" || exit 1

# Switch back to developer user
USER developer

# Set working directory
WORKDIR /workspace

# Entry point
ENTRYPOINT ["/workspace/scripts/startup.sh"]

# Default command
CMD ["/bin/bash"]