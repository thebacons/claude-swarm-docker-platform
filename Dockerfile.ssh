# Claude Swarm Docker Container - SSH Enabled Version
FROM claude-swarm-docker-spawn-policeman:latest

# Switch to root to install SSH
USER root

# Install OpenSSH Server
RUN apt-get update && apt-get install -y \
    openssh-server \
    && rm -rf /var/lib/apt/lists/*

# Configure SSH
RUN mkdir /var/run/sshd && \
    echo 'developer:claude' | chpasswd && \
    sed -i 's/#PermitRootLogin prohibit-password/PermitRootLogin no/' /etc/ssh/sshd_config && \
    sed -i 's/#PasswordAuthentication yes/PasswordAuthentication yes/' /etc/ssh/sshd_config && \
    echo "AllowUsers developer" >> /etc/ssh/sshd_config

# SSH port
EXPOSE 22

# Create SSH startup script
RUN cat > /workspace/scripts/start-ssh.sh << 'EOF'
#!/bin/bash
# Start SSH service
service ssh start

# Run original startup
exec /workspace/scripts/startup.sh "$@"
EOF

RUN chmod +x /workspace/scripts/start-ssh.sh

# Switch back to developer user
USER developer

# Update entrypoint to start SSH
ENTRYPOINT ["/workspace/scripts/start-ssh.sh"]
CMD ["/bin/bash"]