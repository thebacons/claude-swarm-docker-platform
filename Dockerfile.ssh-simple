FROM ubuntu:22.04

# Install SSH and basic tools
RUN apt-get update && apt-get install -y \
    openssh-server \
    sudo \
    curl \
    git \
    python3 \
    python3-pip \
    && rm -rf /var/lib/apt/lists/*

# Create user
RUN useradd -m -s /bin/bash developer && \
    echo "developer:claude"  < /dev/null |  chpasswd && \
    echo "developer ALL=(ALL) NOPASSWD:ALL" >> /etc/sudoers

# Configure SSH
RUN mkdir /var/run/sshd && \
    sed -i 's/#PermitRootLogin prohibit-password/PermitRootLogin no/' /etc/ssh/sshd_config && \
    sed -i 's/#PasswordAuthentication yes/PasswordAuthentication yes/' /etc/ssh/sshd_config

# Create workspace
WORKDIR /workspace
RUN chown developer:developer /workspace

EXPOSE 22

CMD ["/usr/sbin/sshd", "-D"]
