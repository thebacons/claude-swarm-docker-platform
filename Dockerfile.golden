# Claude Code Golden Image - Pre-installed with everything needed
FROM ubuntu:22.04

# Build arguments
ARG CLAUDE_CODE_VERSION="latest"
ARG NODE_VERSION="20"
ARG PYTHON_VERSION="3.11"

# Prevent interactive prompts
ENV DEBIAN_FRONTEND=noninteractive

# Install system dependencies
RUN apt-get update && apt-get install -y \
    curl \
    git \
    wget \
    sudo \
    build-essential \
    software-properties-common \
    ca-certificates \
    gnupg \
    lsb-release \
    jq \
    unzip \
    vim \
    nano \
    htop \
    net-tools \
    iputils-ping \
    && rm -rf /var/lib/apt/lists/*

# Add Python PPA and install Python 3.11
RUN add-apt-repository ppa:deadsnakes/ppa -y && \
    apt-get update && \
    apt-get install -y \
    python${PYTHON_VERSION} \
    python${PYTHON_VERSION}-venv \
    python${PYTHON_VERSION}-dev \
    python${PYTHON_VERSION}-distutils \
    && rm -rf /var/lib/apt/lists/*

# Set Python 3.11 as default
RUN update-alternatives --install /usr/bin/python3 python3 /usr/bin/python${PYTHON_VERSION} 1 && \
    update-alternatives --install /usr/bin/python python /usr/bin/python${PYTHON_VERSION} 1

# Install pip
RUN curl -sS https://bootstrap.pypa.io/get-pip.py | python${PYTHON_VERSION}

# Install Node.js (required for Claude Code and many MCP servers)
RUN curl -fsSL https://deb.nodesource.com/setup_${NODE_VERSION}.x | bash - && \
    apt-get install -y nodejs && \
    npm install -g npm@latest

# Create claude user (matches your local setup)
RUN useradd -m -s /bin/bash claude && \
    echo "claude ALL=(ALL) NOPASSWD:ALL" >> /etc/sudoers

# Set up directory structure
WORKDIR /home/claude
RUN mkdir -p \
    .claude \
    .claude-code \
    .config/claude-code \
    .cache \
    .local/bin \
    .local/share \
    workspace/projects \
    workspace/hooks \
    workspace/scripts \
    workspace/configs \
    workspace/sessions \
    && chown -R claude:claude /home/claude

# Switch to claude user
USER claude
WORKDIR /home/claude

# Install Python packages for orchestration
RUN pip3 install --user \
    anthropic \
    requests \
    pyyaml \
    python-dotenv \
    redis \
    psycopg2-binary \
    aiofiles \
    asyncio \
    rich \
    click

# Copy Claude Code binary and configuration from host
# This assumes you have Claude Code installed on your host machine
COPY --chown=claude:claude --from=host /usr/local/bin/claude /home/claude/.local/bin/claude
COPY --chown=claude:claude --from=host ~/.claude /home/claude/.claude
COPY --chown=claude:claude --from=host ~/.config/claude-code /home/claude/.config/claude-code

# Copy proven hooks system
COPY --chown=claude:claude hooks/ /home/claude/workspace/hooks/

# Copy proven swarm orchestrator
COPY --chown=claude:claude swarm-orchestrator.py /home/claude/workspace/scripts/
COPY --chown=claude:claude demo-swarm.py /home/claude/workspace/scripts/
COPY --chown=claude:claude swarm-config.yaml /home/claude/workspace/scripts/

# Configure environment variables
ENV PATH="/home/claude/.local/bin:${PATH}"
ENV CLAUDE_HOME="/home/claude/.claude"
ENV ANTHROPIC_API_KEY=""

# Create MCP server configuration in workspace root
RUN mkdir -p /home/claude/workspace && \
    cat > /home/claude/workspace/.mcp.json << 'EOF'
{
  "mcpServers": {
    "linear": {
      "transport": "stdio",
      "command": "npx",
      "args": ["-y", "linear-mcp-server"],
      "env": {
        "LINEAR_API_KEY": "${LINEAR_API_KEY}"
      }
    },
    "filesystem": {
      "transport": "stdio",
      "command": "npx",
      "args": ["-y", "@modelcontextprotocol/server-filesystem", "/home/claude/workspace"],
      "env": {}
    },
    "github": {
      "transport": "stdio",
      "command": "npx",
      "args": ["-y", "@modelcontextprotocol/server-github"],
      "env": {
        "GITHUB_PERSONAL_ACCESS_TOKEN": "${GITHUB_PAT_KEY}"
      }
    },
    "ollama": {
      "transport": "http",
      "url": "http://192.168.188.142:11434",
      "headers": {}
    }
  }
}
EOF

# Create Claude Code settings that will load hooks from mounted config
RUN cat > /home/claude/.config/claude-code/settings.json << 'EOF'
{
  "defaultModel": "claude-3-5-sonnet-20241022",
  "apiKey": "${ANTHROPIC_API_KEY}",
  "hooksConfig": "/home/claude/workspace/hooks/config/claude-hooks.json"
}
EOF

# Create startup script that initializes Claude Code
RUN cat > /home/claude/startup.sh << 'EOF'
#!/bin/bash

echo "=== CASPER Golden Image Starting ==="
echo "Claude Code Pre-installed Edition"
echo ""

# Load environment from .env if mounted
if [ -f /home/claude/workspace/.env ]; then
    echo "Loading environment variables..."
    set -a
    source /home/claude/workspace/.env
    set +a
fi

# Export critical environment variables
export ANTHROPIC_API_KEY="${ANTHROPIC_API_KEY}"
export GITHUB_PAT_KEY="${GITHUB_PAT_KEY}"
export LINEAR_API_KEY="${LINEAR_API_KEY}"
export AGENT_ROLE="${AGENT_ROLE:-orchestrator}"

# Verify Claude Code installation
if command -v claude &> /dev/null; then
    echo "✅ Claude Code is installed"
    claude --version || echo "Version check not available"
else
    echo "❌ Claude Code not found!"
fi

# Test MCP servers
echo ""
echo "Testing MCP server connectivity..."
claude mcp list 2>/dev/null || echo "MCP server list not available"

# Check hook configuration
echo ""
echo "Hook system status:"
if [ -d /home/claude/workspace/hooks ]; then
    echo "✅ Hooks directory mounted"
    # Make all hook scripts executable
    find /home/claude/workspace/hooks -name "*.sh" -exec chmod +x {} \; 2>/dev/null
    echo "✅ Hook scripts made executable"
    
    # Load hook configuration if available
    if [ -f /home/claude/workspace/hooks/config/claude-hooks.json ]; then
        echo "✅ Hook configuration found"
        # Copy to Claude Code settings location if needed
        mkdir -p /home/claude/.config/claude-code
        cp /home/claude/workspace/hooks/config/claude-hooks.json /home/claude/.config/claude-code/hooks.json
    fi
else
    echo "⚠️  Hooks directory not found"
fi

# Agent role-specific initialization
case "$AGENT_ROLE" in
    "orchestrator"|"policeman")
        echo ""
        echo "=== Policeman Orchestrator Mode ==="
        echo "You are the master orchestrator with ability to:"
        echo "- Spawn and coordinate multiple agents"
        echo "- Use proven swarm orchestration"
        echo "- Validate all outputs with hooks"
        echo ""
        echo "To start orchestration:"
        echo "  claude 'Orchestrate: [your task]'"
        echo ""
        echo "Or use direct swarm control:"
        echo "  python3 /home/claude/workspace/scripts/swarm-orchestrator.py"
        ;;
    "developer")
        echo ""
        echo "=== Developer Agent Mode ==="
        echo "Specialization: ${AGENT_SPECIALIZATION:-general}"
        echo "Ready to receive tasks from orchestrator"
        ;;
    "tester")
        echo ""
        echo "=== Tester Agent Mode ==="
        echo "Ready to validate code quality"
        ;;
esac

# Start in workspace
cd /home/claude/workspace

echo ""
echo "=== Golden Image Ready ==="
echo ""

# If interactive shell, start bash
if [ -t 0 ]; then
    exec /bin/bash
else
    # Keep container running for orchestration
    tail -f /dev/null
fi
EOF

RUN chmod +x /home/claude/startup.sh

# Create orchestrator wrapper that understands its role
RUN cat > /home/claude/.local/bin/casper-orchestrate << 'EOF'
#!/bin/bash

# CASPER Orchestrator - Launches Claude Code with orchestrator context

SYSTEM_PROMPT="You are CASPER's Policeman Orchestrator running in a Docker container.

You have the ability to:
1. Spawn and coordinate multiple AI agents (up to 15 in parallel)
2. Each agent runs in its own container (developer-1, developer-2, tester)
3. Use the proven swarm orchestration at /home/claude/workspace/scripts/swarm-orchestrator.py
4. All outputs are validated by hooks before acceptance
5. You can spawn additional instances using docker-compose scale

For parallel tasks, use:
  python3 /home/claude/workspace/scripts/swarm-orchestrator.py

For performance testing:
  python3 /home/claude/workspace/scripts/demo-swarm.py

Remember: You achieved 3.7x performance with parallel execution. Use it!"

# Launch Claude Code with orchestrator context
claude --system-prompt "$SYSTEM_PROMPT" "$@"
EOF

RUN chmod +x /home/claude/.local/bin/casper-orchestrate

# Set up health check
HEALTHCHECK --interval=30s --timeout=10s --start-period=5s --retries=3 \
    CMD claude --version || exit 1

# Entry point
ENTRYPOINT ["/home/claude/startup.sh"]

# Default command
CMD ["/bin/bash"]